### PR TITLE
Prml 447 infered models in train new

### PR DESCRIPTION
**Pull request by subham-71**
## Description of the change

> Description here

## Type of change
- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] The code changed/added as part of this pull request has been covered with tests
- [ ] All tests related to the changed code pass in development

### Code review 

- [ ]  This pull request has a descriptive title and information useful to a reviewer. There may be a screenshot or screencast attached
- [ ] "Ready for review" label attached to the PR and reviewers mentioned in a comment
- [ ] Changes have been reviewed by at least one other engineer
- [ ] Issue from task tracker has a link to this pull request

### CODE REVIEW

File: samples/application_project/base_features/models/profiles-ml.yaml [hunk 1]
```diff
@@ -0,0 +1,46 @@
+models:
+  - name: &model_name shopify_churn_base
```
Comment (shekhar-rudder) [inline samples/application_project/base_features/models/profiles-ml.yaml:2]: Do we need python model in both base and application ? I think we can keep only the base one and remove it from application

File: samples/application_project/base_features/models/profiles-ml.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    model_type: python_model
+    model_spec:
+      occurred_at_col: insert_ts
+      entity_key: user
+      validity_time: 0s # 1 day
+      py_repo_url: /Users/admin/Desktop/Profiles/rudderstack_profiles_classifier
```
Comment (shekhar-rudder) [inline samples/application_project/base_features/models/profiles-ml.yaml:8]: Use relative path since this exact path won't exist for all the devs

File: src/predictions/rudderstack_predictions/train.py [hunk 3]
```diff
@@ -88,8 +88,13 @@ def _train(
     _ = config["data"].pop(
         "features_profiles_model", None
     )  # For backward compatibility. Not using it anywhere else, hence deleting.
+
     merged_config = utils.combine_config(default_config, config)
 
+    _ = merged_config["data"].pop(
```
Comment (shekhar-rudder) [inline src/predictions/rudderstack_predictions/train.py:94]: `inputs` should be popped from `config` instead of `merged_config` for consistency with the rest of code

File: src/predictions/rudderstack_predictions/train.py [hunk 4]
```diff
@@ -274,6 +279,8 @@ def train_and_store_model_results_sf(
         )
         trainer.label_value = label_value
 
+    new_input_models = whtService.get_input_models(inputs, output_filename)
```
Comment (shekhar-rudder) [inline src/predictions/rudderstack_predictions/train.py:282]: There is no reason to call the result as "new". Lets rename it to `input_models`

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py [hunk 5]
```diff
@@ -38,6 +40,49 @@ def _getPB(self):
...
+        }
+
+        pb_show_models_response_output = self._getPB().show_models(args)
+
+        # Find matches for input models in the JSON data
+        new_input_models = []
+
+        if len(original_input_models) != 0:
+            for model in original_input_models:
+                model_key = model.split("/")[-1]
```
Comment (shekhar-rudder) [inline src/predictions/rudderstack_predictions/wht/pythonWHT.py:72]: I'm not sure how this is working. `original_input_models` is supposed to be a list of model names and names don't contain "/". Can you add some unit tests for this function

File: src/predictions/rudderstack_predictions/train.py (hunk 3 above)
Comment (dpatchigolla) [reply src/predictions/rudderstack_predictions/train.py:95]: In addition to this, `inputs` key should be removed from the model_config.yaml file.

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py [hunk 6]
```diff
@@ -38,6 +40,49 @@ def _getPB(self):
...
         return RudderPB()
 
+    def get_input_models(
+        self,
+        inputs: List[str],
+        train_summary_output_file_name: str,
+    ) -> List[str]:
+        """Find matches for input models in the JSON data. If no matches are found, an exception is raised.
+        Args:
+            inputs (List[str]): List of inputs
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/wht/pythonWHT.py:50]: `inputs (List[str]): List of inputs` to:
`inputs (List[str]): List of inputs - sql query strings generated by profiles and sent as params in train()`

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py (hunk 5 above)
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/wht/pythonWHT.py:72]: From line 57, original_input_models seem to be generated from splitting material names, so where does it have the `/` character?

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py [hunk 7]
```diff
@@ -38,6 +40,49 @@ def _getPB(self):
...
+            "site_config_path": self.site_config_path,
+            "project_folder": self.project_folder_path,
+        }
+
+        pb_show_models_response_output = self._getPB().show_models(args)
+
+        # Find matches for input models in the JSON data
+        new_input_models = []
+
+        if len(original_input_models) != 0:
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/wht/pythonWHT.py:70]: Why is this check required?

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py [hunk 8]
```diff
@@ -38,6 +40,49 @@ def _getPB(self):
...
+            for model in original_input_models:
+                model_key = model.split("/")[-1]
+                found_unique_match = False
+                for key in pb_show_models_response_output:
+                    if key.endswith(model_key):
+                        if found_unique_match:
+                            raise ValueError(
+                                f"Multiple unique occurrences found for {model_key}"
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/wht/pythonWHT.py:78]: f"Multiple unique occurrences found for {model_key}" Change this to:

f"Multiple models with name {model_key} are found. Please ensure the models added in inputs are named uniquely and retry"

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py [hunk 9]
```diff
@@ -38,6 +40,49 @@ def _getPB(self):
...
+        pb_show_models_response_output = self._getPB().show_models(args)
+
+        # Find matches for input models in the JSON data
+        new_input_models = []
+
+        if len(original_input_models) != 0:
+            for model in original_input_models:
+                model_key = model.split("/")[-1]
+                found_unique_match = False
+                for key in pb_show_models_response_output:
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/wht/pythonWHT.py:74]:
```    for model_key in model_keys:
        matching_keys = [key for key in pb_show_models_response_output if key.endswith(model_key)]
        if len(matching_keys) > 1:
            raise ValueError(f"Multiple unique occurrences found for {model_key}")
        elif matching_keys:
            new_input_models.append(matching_keys[0].split("/", 1)[-1])
```

Can you structure your code to something like this? fewer nested if conditions and for-loops making it look more concise? 
Check for code correctness.

File: src/predictions/rudderstack_predictions/wht/rudderPB.py [hunk 10]
```diff
@@ -86,16 +86,26 @@ def show_models(self, arg: dict) -> List[str]:
...
-        )
-        if start_index is None:
+        # Find the index of the first '{' after the line
+        start_index = stdout.find("{", start_index)
+
+        if start_index == -1:
             return None
 
+        # Find the index of the last '}'
+        end_index = stdout.rfind("}")
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/wht/rudderPB.py:100]: You had a unit test to cover this function in your prev pr right? Can you add that back?

**Review by dpatchigolla — COMMENTED** [review]
Looks good overall. Minor comments provided inline. Please address all the comments.

File: samples/application_project/base_features/models/profiles-ml.yaml (hunk 1 above)
Comment (subham-71) [reply samples/application_project/base_features/models/profiles-ml.yaml:2]: We need in both

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py (hunk 5 above)
Comment (subham-71) [reply src/predictions/rudderstack_predictions/wht/pythonWHT.py]: We earlier needed this when we weren't using the split_material_name function. We don't need this anymore.

File: src/predictions/rudderstack_predictions/wht/rudderPB.py (hunk 10 above)
Comment (subham-71) [reply src/predictions/rudderstack_predictions/wht/rudderPB.py:100]: I have added the tests

File: src/predictions/rudderstack_predictions/train.py [hunk 11]
```diff
@@ -275,6 +279,8 @@ def train_and_store_model_results_sf(
         )
         trainer.label_value = label_value
 
+    new_input_models = whtService.get_input_models(inputs)
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/train.py:282]: This needs to be renamed to `absolute_input_models` as per Shashank's comment earlier.

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py [hunk 12]
```diff
@@ -43,45 +43,45 @@ def _getPB(self):
...
-            inputs (List[str]): List of inputs
+            inputs (List[str]): List of inputs - sql query strings generated by profiles and sent as params in train()
             train_summary_output_file_name (str): output filename
         Returns:
             List[str]: List of input models - full paths in the profiles project for models that are required to generate the current model.
         """
-        # original_input_models : List of input models - relative paths in the profiles project for models that are required to generate the current model.
         original_input_models = [
-            self._split_material_name(input_)["model_name"] for input_ in inputs
+            self.split_material_name(input_)["model_name"] for input_ in inputs
```
Comment (dpatchigolla) [inline src/predictions/rudderstack_predictions/wht/pythonWHT.py:55]: Why did you remove the `_` prefix? Is this function being used anywhere else?

**Review by dpatchigolla — COMMENTED** [review]
Only one rename suggestion.

File: src/predictions/rudderstack_predictions/wht/pythonWHT.py (hunk 12 above)
Comment (subham-71) [reply src/predictions/rudderstack_predictions/wht/pythonWHT.py:55]: This is because, the function was updated later from  _split_material_name to split_material_name.

**Review by dpatchigolla — APPROVED** [review]
